src/js/modules/utilities/IOBuf.ts

```diff
@@ -206,19 +226,43 @@ export class IOBuf {
     return iob;
   }
 
+  /**
+   * find the first fragment with free space, otherwise it returns the last
+   * fragment in the list.
+   */
   private getLastFragment = (): Fragment => {
-    const lastFragment = this.fragmentList[this.fragmentList.length - 1];
+    const lastFragment = this.fragmentList[this.currentWriteIndex];
     if (lastFragment) {
       return lastFragment;
     } else {
-      return this.appendNewFragment();
+      return this.appendNewFragment(lastFragment?.getSize());
+    }
+  };
+
+  /**
+   * given a fragment, it tries to find the next fragment in the list , if there
+   * isn't next one, it appends new one.
+   * @param fragment
+   */
+  private getNextFragmentOrAppend = (fragment: Fragment): Fragment => {
+    const nextFragment = this.fragmentList[this.currentWriteIndex + 1];
+    if (nextFragment) {
+      return nextFragment;
+    } else {
+      return this.appendNewFragment(fragment.getSize());
     }
   };
 
   private appendNewFragment = (size?: number): Fragment => {
     const nextSize = size ? getNextSize(size) : fragmentSizes[0];
     const fragment = new Fragment(nextSize);
-    this.fragmentList.push(fragment);
+    const newIndex = this.fragmentList.push(fragment);
+    // if newIndex is 1, its mean this.fragment added the first element, in this
+    // case, the currentIndex shouldn't , 0 + 1 = 1, 1 can't be the
+    // first index, array start on 0
```

Comment:
I don't understand this part :(

Reply:
I changed it, let me know if it is better